### Data Extract
- Manual operation through Catalog > Create Table

### Bronze Layer

In [0]:
%sql
CREATE DATABASE IF NOT EXISTS case_study;
USE case_study;


CREATE OR REPLACE TEMPORARY VIEW raw_products
USING CSV
OPTIONS (
  path '/FileStore/tables/products__1_.csv',
  header 'true',
  inferSchema 'true'
);

CREATE OR REPLACE TEMPORARY VIEW raw_sales_order_detail
USING CSV
OPTIONS (
  path '/FileStore/tables/sales_order_detail__1_.csv',
  header 'true',
  inferSchema 'true'
);

CREATE OR REPLACE TEMPORARY VIEW raw_sales_order_header
USING CSV
OPTIONS (
  path '/FileStore/tables/sales_order_header__1_.csv',
  header 'true',
  inferSchema 'true'
);

In [0]:
%sql

SELECT 'raw_products' as table_name, COUNT(*) as row_count FROM raw_products
UNION ALL
SELECT 'raw_sales_order_detail', COUNT(*) FROM raw_sales_order_detail
UNION ALL
SELECT 'raw_sales_order_header', COUNT(*) FROM raw_sales_order_header;

table_name row_count raw_products 303 raw_sales_order_detail 121317 raw_sales_order_header 31465

### Check DataTypes

In [0]:
%sql
DESCRIBE raw_products;

col_name data_type comment ProductID int null ProductDesc string null ProductNumber string null MakeFlag boolean null Color string null SafetyStockLevel int null ReorderPoint int null StandardCost double null ListPrice double null Size string null SizeUnitMeasureCode string null Weight double null WeightUnitMeasureCode string null ProductCategoryName string null ProductSubCategoryName string null

In [0]:
%sql
DESCRIBE raw_sales_order_detail;

col_name data_type comment SalesOrderID int null SalesOrderDetailID int null OrderQty int null ProductID int null UnitPrice double null UnitPriceDiscount double null

In [0]:
%sql
DESCRIBE raw_sales_order_header;

col_name data_type comment SalesOrderID int null OrderDate timestamp null ShipDate date null OnlineOrderFlag boolean null AccountNumber string null CustomerID int null SalesPersonID int null Freight double null

### Create Store Tables With Proper DataTypes
- Data Quality Check

### Silver Layer

In [0]:
from pyspark.sql.functions import col
from pyspark.sql.types import IntegerType, StringType, BooleanType, DecimalType

# Read raw_products
df_raw = spark.table("raw_products")

# Apply all type conversions
df_typed = df_raw.select(
    col("ProductID").cast(IntegerType()).alias("ProductID"),
    col("ProductDesc").cast(StringType()).alias("ProductDesc"),
    col("ProductNumber").cast(StringType()).alias("ProductNumber"),
    col("MakeFlag").cast(BooleanType()).alias("MakeFlag"),
    col("Color").cast(StringType()).alias("Color"),
    col("SafetyStockLevel").cast(IntegerType()).alias("SafetyStockLevel"),
    col("ReorderPoint").cast(IntegerType()).alias("ReorderPoint"),
    col("StandardCost").cast(DecimalType(10, 2)).alias("StandardCost"),
    col("ListPrice").cast(DecimalType(10, 2)).alias("ListPrice"),
    col("Size").cast(StringType()).alias("Size"),
    col("SizeUnitMeasureCode").cast(StringType()).alias("SizeUnitMeasureCode"),
    col("Weight").cast(DecimalType(10, 2)).alias("Weight"),
    col("WeightUnitMeasureCode").cast(StringType()).alias("WeightUnitMeasureCode"),
    col("ProductCategoryName").cast(StringType()).alias("ProductCategoryName"),
    col("ProductSubCategoryName").cast(StringType()).alias("ProductSubCategoryName")
)
# Write as table (overwrite if exists)
df_typed.write.mode("overwrite").saveAsTable("store_products")
# Count verification
product_count = spark.table("store_products").count()
print(f"Product count: {product_count}")

Product count: 303


In [0]:
%sql
CREATE OR REPLACE TABLE store_sales_order_detail AS
SELECT 
    CAST(SalesOrderID AS INT) AS SalesOrderID,           -- Foreign Key to header
    CAST(SalesOrderDetailID AS INT) AS SalesOrderDetailID, -- Primary Key
    CAST(OrderQty AS INT) AS OrderQty,
    CAST(ProductID AS INT) AS ProductID,                 -- Foreign Key to products
    CAST(UnitPrice AS DECIMAL(10,2)) AS UnitPrice,       -- Money type
    CAST(UnitPriceDiscount AS DECIMAL(10,2)) AS UnitPriceDiscount  -- Money type
FROM raw_sales_order_detail;

-- Count verification
SELECT COUNT(*) as detail_count FROM store_sales_order_detail;

detail_count 121317

In [0]:
%sql
CREATE OR REPLACE TABLE store_sales_order_header AS
SELECT 
    CAST(SalesOrderID AS INT) AS SalesOrderID,           -- Primary Key
    CAST(OrderDate AS DATE) AS OrderDate,                -- Ensure DATE type
    CAST(ShipDate AS DATE) AS ShipDate,                  -- Ensure DATE type
    CAST(OnlineOrderFlag AS BOOLEAN) AS OnlineOrderFlag,
    CAST(AccountNumber AS STRING) AS AccountNumber,
    CAST(CustomerID AS INT) AS CustomerID,               -- Foreign Key (to customer table)
    CAST(SalesPersonID AS INT) AS SalesPersonID,
    CAST(Freight AS DECIMAL(10,2)) AS Freight            -- Money type
FROM raw_sales_order_header;

-- Count verification
SELECT COUNT(*) as header_count FROM store_sales_order_header;

header_count 31465

### Gold Layer

In [0]:
%sql

CREATE OR REPLACE TABLE publish_orders AS
SELECT 
    -- All fields from SalesOrderDetail
    d.SalesOrderID,
    d.SalesOrderDetailID,
    d.OrderQty,
    d.ProductID,
    d.UnitPrice,
    d.UnitPriceDiscount,
    
    -- All fields from SalesOrderHeader (except SalesOrderID which is already included)
    h.OrderDate,
    h.ShipDate,
    h.OnlineOrderFlag,
    h.AccountNumber,
    h.CustomerID,
    h.SalesPersonID,
    h.Freight AS TotalOrderFreight,  -- Renamed as required
    
    -- Calculated fields
    calculate_business_days(h.OrderDate, h.ShipDate) AS LeadTimeInBusinessDays,
    d.OrderQty * (d.UnitPrice - d.UnitPriceDiscount) AS TotalLineExtendedPrice
    
FROM store_sales_order_detail d
INNER JOIN store_sales_order_header h 
    ON d.SalesOrderID = h.SalesOrderID;

-- Verify the table was created
SELECT COUNT(*) as total_orders FROM publish_orders;

total_orders 121317

In [0]:
    %sql
CREATE OR REPLACE TABLE publish_product AS
SELECT 
    ProductID,
    ProductDesc,
    ProductNumber,
    MakeFlag,
    -- Replace NULL values in Color with 'N/A'
    COALESCE(Color, 'N/A') AS Color,
    SafetyStockLevel,
    ReorderPoint,
    StandardCost,
    ListPrice,
    Size,
    SizeUnitMeasureCode,
    Weight,
    WeightUnitMeasureCode,
    -- Enhance ProductCategoryName when NULL
    CASE 
        WHEN ProductCategoryName IS NOT NULL THEN ProductCategoryName
        WHEN ProductSubCategoryName IN ('Gloves', 'Shorts', 'Socks', 'Tights', 'Vests') THEN 'Clothing'
        WHEN ProductSubCategoryName IN ('Locks', 'Lights', 'Headsets', 'Helmets', 'Pedals', 'Pumps') THEN 'Accessories'
        WHEN ProductSubCategoryName LIKE '%Frames%' OR ProductSubCategoryName IN ('Wheels', 'Saddles') THEN 'Components'
        ELSE ProductCategoryName
    END AS ProductCategoryName,
    ProductSubCategoryName
FROM store_products;

-- Verify the transformations
SELECT 
    'Total Products' as metric, COUNT(*) as value 
FROM publish_product
UNION ALL
SELECT 
    'Products with N/A Color', COUNT(*) 
FROM publish_product 
WHERE Color = 'N/A'
UNION ALL
SELECT 
    'NULL Categories after transformation', COUNT(*) 
FROM publish_product 
WHERE ProductCategoryName IS NULL;

metric value Products with N/A Color 50 NULL Categories after transformation 46 Total Products 303

### Transformations

In [0]:
%sql
-- Check BEFORE transformation (only store_products)
SELECT 
    'BEFORE - NULL Categories in store_products' as metric, COUNT(*) as value
FROM store_products WHERE ProductCategoryName IS NULL
UNION ALL
SELECT 
    'BEFORE - NULL Colors in store_products', COUNT(*)
FROM store_products WHERE Color IS NULL;

metric value BEFORE - NULL Categories in store_products 190 BEFORE - NULL Colors in store_products 50

In [0]:
%sql
-- Check AFTER transformation
SELECT 
    'AFTER - NULL Categories in publish_product' as metric, COUNT(*) as value
FROM publish_product WHERE ProductCategoryName IS NULL
UNION ALL
SELECT 
    'AFTER - Products with N/A Color', COUNT(*)
FROM publish_product WHERE Color = 'N/A';

metric value AFTER - NULL Categories in publish_product 46 AFTER - Products with N/A Color 50

In [0]:
%sql
-- Check data quality BEFORE transformation in store_products
SELECT 
    'BEFORE - NULL Categories in store_products' as metric, 
    COUNT(*) as value,
    ROUND(COUNT(*) * 100.0 / 303, 1) as percentage
FROM store_products 
WHERE ProductCategoryName IS NULL

UNION ALL

SELECT 
    'BEFORE - NULL Colors in store_products', 
    COUNT(*),
    ROUND(COUNT(*) * 100.0 / 303, 1)
FROM store_products 
WHERE Color IS NULL

UNION ALL

-- Also check distribution of categories before transformation
SELECT 
    'BEFORE - Products WITH Category', 
    COUNT(*),
    ROUND(COUNT(*) * 100.0 / 303, 1)
FROM store_products 
WHERE ProductCategoryName IS NOT NULL;

-- Show sample of NULL category products
SELECT ProductID, ProductDesc, ProductSubCategoryName, ProductCategoryName
FROM store_products
WHERE ProductCategoryName IS NULL
LIMIT 5;

ProductID ProductDesc ProductSubCategoryName ProductCategoryName 680 HL Road Frame - Black, 58 Road Frames null 706 HL Road Frame - Red, 58 Road Frames null 707 Sport-100 Helmet, Red Helmets null 708 Sport-100 Helmet, Black Helmets null 709 Mountain Bike Socks, M Socks null

In [0]:
%sql

SELECT 
    COALESCE(ProductCategoryName, 'NULL') as Category, 
    COUNT(*) as ProductCount
FROM publish_product
GROUP BY ProductCategoryName
ORDER BY ProductCount DESC;

Category ProductCount Components 145 Clothing 53 NULL 46 Bikes 32 Accessories 27

In [0]:
%sql
CREATE OR REPLACE FUNCTION calculate_business_days(start_date DATE, end_date DATE)
RETURNS INT
RETURN 
  CASE 
    WHEN start_date IS NULL OR end_date IS NULL THEN NULL
    ELSE 
      -- Total days minus weekends
      DATEDIFF(end_date, start_date) 
      -- Subtract complete weeks * 2 (Saturday and Sunday)
      - (2 * FLOOR(DATEDIFF(end_date, start_date) / 7))
      -- Adjust for partial week
      - CASE 
          -- If start is Sunday and end is not Sunday, subtract 1
          WHEN DAYOFWEEK(start_date) = 1 AND DAYOFWEEK(end_date) != 1 THEN 1
          -- If start is not Saturday and end is Saturday, subtract 1
          WHEN DAYOFWEEK(start_date) != 7 AND DAYOFWEEK(end_date) = 7 THEN 1
          -- If start is Sunday and end is Saturday, subtract 2
          WHEN DAYOFWEEK(start_date) = 1 AND DAYOFWEEK(end_date) = 7 THEN 2
          ELSE 0
        END
  END;

In [0]:
%sql
SELECT COUNT(*) as null_lead_time_count 
FROM publish_orders 
WHERE LeadTimeInBusinessDays IS NULL;

SELECT * 
FROM publish_orders 
WHERE LeadTimeInBusinessDays > 20
LIMIT 10;

SELECT 
    OrderQty, 
    UnitPrice, 
    UnitPriceDiscount,
    TotalLineExtendedPrice,
    OrderQty * (UnitPrice - UnitPriceDiscount) as CheckCalc,
    CASE 
        WHEN TotalLineExtendedPrice = OrderQty * (UnitPrice - UnitPriceDiscount) 
        THEN 'OK' 
        ELSE 'ERROR' 
    END as Validation
FROM publish_orders 
LIMIT 10;

OrderQty UnitPrice UnitPriceDiscount TotalLineExtendedPrice CheckCalc Validation 1 2024.99 0.00 2024.99 2024.99 OK 3 2024.99 0.00 6074.97 6074.97 OK 1 2024.99 0.00 2024.99 2024.99 OK 1 2039.99 0.00 2039.99 2039.99 OK 1 2039.99 0.00 2039.99 2039.99 OK 2 2039.99 0.00 4079.98 4079.98 OK 1 2039.99 0.00 2039.99 2039.99 OK 3 28.84 0.00 86.52 86.52 OK 1 28.84 0.00 28.84 28.84 OK 6 5.70 0.00 34.20 34.20 OK

In [0]:
%sql

SELECT 
    SalesOrderID,
    OrderDate,
    ShipDate,
    LeadTimeInBusinessDays,
    OrderQty,
    UnitPrice,
    UnitPriceDiscount,
    TotalLineExtendedPrice
FROM publish_orders
LIMIT 10;

SalesOrderID OrderDate ShipDate LeadTimeInBusinessDays OrderQty UnitPrice UnitPriceDiscount TotalLineExtendedPrice 43659 2021-05-31 2021-06-07 5 1 2024.99 0.00 2024.99 43659 2021-05-31 2021-06-07 5 3 2024.99 0.00 6074.97 43659 2021-05-31 2021-06-07 5 1 2024.99 0.00 2024.99 43659 2021-05-31 2021-06-07 5 1 2039.99 0.00 2039.99 43659 2021-05-31 2021-06-07 5 1 2039.99 0.00 2039.99 43659 2021-05-31 2021-06-07 5 2 2039.99 0.00 4079.98 43659 2021-05-31 2021-06-07 5 1 2039.99 0.00 2039.99 43659 2021-05-31 2021-06-07 5 3 28.84 0.00 86.52 43659 2021-05-31 2021-06-07 5 1 28.84 0.00 28.84 43659 2021-05-31 2021-06-07 5 6 5.70 0.00 34.20

### Measure Performance

In [0]:
import time
# Test 1: Simple count query
start_time = time.time()
count = spark.sql("SELECT COUNT(*) FROM publish_orders").collect()[0][0]
end_time = time.time()
print(f"1. Count query time: {end_time - start_time:.2f} seconds")
print(f"   Total records: {count:,}")
# Test 2: Aggregation query (revenue by year)
start_time = time.time()
spark.sql("""
    SELECT YEAR(OrderDate) as Year, SUM(TotalLineExtendedPrice) as Revenue
    FROM publish_orders
    GROUP BY YEAR(OrderDate)
""").collect()
end_time = time.time()
print(f"\n2. Aggregation query time: {end_time - start_time:.2f} seconds")
# Test 3: Join query performance
start_time = time.time()
spark.sql("""
    SELECT COUNT(*)
    FROM publish_orders o
    JOIN publish_product p ON o.ProductID = p.ProductID
""").collect()
end_time = time.time()
print(f"\n3. Join query time: {end_time - start_time:.2f} seconds")
# Test 4: Complex analysis query
start_time = time.time()
spark.sql("""
    SELECT p.ProductCategoryName, COUNT(*) as Orders
    FROM publish_orders o
    JOIN publish_product p ON o.ProductID = p.ProductID
    GROUP BY p.ProductCategoryName
""").collect()
end_time = time.time()
print(f"\n4. Complex analysis time: {end_time - start_time:.2f} seconds")
# Storage information
print("\n5. Storage Information:")
spark.sql("DESCRIBE DETAIL publish_orders").select("sizeInBytes", "numFiles").show()
spark.sql("DESCRIBE DETAIL publish_product").select("sizeInBytes", "numFiles").show()

1. Count query time: 0.99 seconds
   Total records: 121,317

2. Aggregation query time: 1.05 seconds

3. Join query time: 1.34 seconds

4. Complex analysis time: 1.41 seconds

5. Storage Information:
+-----------+--------+
|sizeInBytes|numFiles|
+-----------+--------+
|    1961779|       2|
+-----------+--------+

+-----------+--------+
|sizeInBytes|numFiles|
+-----------+--------+
|      13630|       1|
+-----------+--------+



In [0]:

# 1. Check file paths
try:
    display(dbutils.fs.ls("/FileStore/tables/"))
    print("✓ File paths accessible")
except:
    print("✗ Issue with file paths - use dbutils.fs.ls() to verify")

# 2. Check for non-numeric IDs (potential CAST errors)
invalid_ids = spark.sql("""
    SELECT COUNT(*) as invalid_count
    FROM raw_products
    WHERE ProductID IS NULL 
       OR NOT CAST(ProductID AS STRING) RLIKE '^[0-9]+$'
""").collect()[0][0]
print(f"\n2. Invalid ProductIDs: {invalid_ids}")

# 3. Check for special characters in money fields
money_issues = spark.sql("""
    SELECT COUNT(*) as issues
    FROM raw_products
    WHERE StandardCost LIKE '%$%' 
       OR StandardCost LIKE '%,%'
""").collect()[0][0]
print(f"3. Money format issues: {money_issues}")

# 4. Common Databricks Community limitations
print("\n=== PLATFORM LIMITATIONS ===")
print("- No PRIMARY/FOREIGN KEY support")
print("- No Unity Catalog features")
print("- Limited cluster resources")
print("- Session timeout after 2 hours")

path name size modificationTime dbfs:/FileStore/tables/10Y_lake-1.xlsx 10Y_lake-1.xlsx 124274 1667071700000 dbfs:/FileStore/tables/10Y_lake.xlsx 10Y_lake.xlsx 124274 1667071572000 dbfs:/FileStore/tables/GRD_PUBLICO_2019_formatted-1.txt GRD_PUBLICO_2019_formatted-1.txt 572538020 1742486023000 dbfs:/FileStore/tables/GRD_PUBLICO_2019_formatted.txt GRD_PUBLICO_2019_formatted.txt 572538020 1742440723000 dbfs:/FileStore/tables/GRD_PUBLICO_2020_formated_csv.xlsx GRD_PUBLICO_2020_formated_csv.xlsx 212379385 1742436082000 dbfs:/FileStore/tables/GRD_PUBLICO_2020_formated_csvformat.csv GRD_PUBLICO_2020_formated_csvformat.csv 399887780 1742437578000 dbfs:/FileStore/tables/GRD_PUBLICO_2020_formatted.txt GRD_PUBLICO_2020_formatted.txt 397026707 1742486224000 dbfs:/FileStore/tables/GRD_PUBLICO_2021_csv.csv GRD_PUBLICO_2021_csv.csv 427870918 1742439551000 dbfs:/FileStore/tables/GRD_PUBLICO_2021_formatted.txt GRD_PUBLICO_2021_formatted.txt 428657825 1742486403000 dbfs:/FileStore/tables/GRD_PUBLICO_2023_csvformatted.csv GRD_PUBLICO_2023_csvformatted.csv 541339028 1742438024000 dbfs:/FileStore/tables/GRD_PUBLICO_EXTERNO_2022.txt GRD_PUBLICO_EXTERNO_2022.txt 960774530 1742487314000 dbfs:/FileStore/tables/GRD_PUBLICO_EXTERNO_2022_csv.csv GRD_PUBLICO_EXTERNO_2022_csv.csv 484639408 1742439042000 dbfs:/FileStore/tables/GRD_PUBLICO_EXTERNO_2022_formatted.txt GRD_PUBLICO_EXTERNO_2022_formatted.txt 960774530 1742486714000 dbfs:/FileStore/tables/Petrol_Dataset_June_20_2022.csv Petrol_Dataset_June_20_2022.csv 8937 1656268306000 dbfs:/FileStore/tables/Test-1.csv Test-1.csv 527709 1658604017000 dbfs:/FileStore/tables/Test.csv Test.csv 527709 1658596960000 dbfs:/FileStore/tables/alura_dataset.csv alura_dataset.csv 22597 1662758886000 dbfs:/FileStore/tables/grd_publico_2020_csv.csv grd_publico_2020_csv.csv 399887780 1742440229000 dbfs:/FileStore/tables/movies-1.csv movies-1.csv 1352598 1658603330000 dbfs:/FileStore/tables/movies-2.csv movies-2.csv 1352598 1658603587000 dbfs:/FileStore/tables/movies.csv movies.csv 1352598 1658596489000 dbfs:/FileStore/tables/products__1_.csv products__1_.csv 39068 1753463178000 dbfs:/FileStore/tables/sales_order_detail__1_-1.csv sales_order_detail__1_-1.csv 5521253 1753463360000 dbfs:/FileStore/tables/sales_order_detail__1_.csv sales_order_detail__1_.csv 5521253 1753463242000 dbfs:/FileStore/tables/sales_order_header__1_.csv sales_order_header__1_.csv 2517409 1753463289000 dbfs:/FileStore/tables/tabela_bagunca.csv tabela_bagunca.csv 103884 1667071819000 dbfs:/FileStore/tables/teste_csv-1.csv teste_csv-1.csv 6485 1668297784000 dbfs:/FileStore/tables/teste_csv-2.csv teste_csv-2.csv 6485 1668297841000 dbfs:/FileStore/tables/teste_csv.csv teste_csv.csv 6485 1668297675000

✓ File paths accessible

2. Invalid ProductIDs: 0
3. Money format issues: 0

=== PLATFORM LIMITATIONS ===
- No PRIMARY/FOREIGN KEY support
- No Unity Catalog features
- Limited cluster resources
- Session timeout after 2 hours


### Analysis 1
- Highest revenue by color per year

In [0]:
%sql
WITH revenue_by_year_color AS (
  SELECT 
    YEAR(o.OrderDate) AS Year,
    p.Color,
    SUM(o.TotalLineExtendedPrice) AS TotalRevenue
  FROM publish_orders o
  JOIN publish_product p ON o.ProductID = p.ProductID
  WHERE o.OrderDate IS NOT NULL
  GROUP BY YEAR(o.OrderDate), p.Color
),
ranked_colors AS (
  SELECT 
    Year,
    Color,
    TotalRevenue,
    ROW_NUMBER() OVER (PARTITION BY Year ORDER BY TotalRevenue DESC) as rank
  FROM revenue_by_year_color
)
SELECT 
  Year,
  Color,
  FORMAT_NUMBER(TotalRevenue, 2) AS TotalRevenue
FROM ranked_colors
WHERE rank = 1
ORDER BY Year;

Year Color TotalRevenue 2021 Red 6,019,613.15 2022 Black 14,005,216.19 2023 Black 15,047,626.43 2024 Yellow 6,480,720.07

### Analysis 2
- What is the average LeadTimeInBusinessDays by ProductCategoryName?

In [0]:
%sql
SELECT 
  p.ProductCategoryName,
  ROUND(AVG(o.LeadTimeInBusinessDays), 2) AS AvgLeadTimeInBusinessDays,
  COUNT(*) AS TotalOrders,
  MIN(o.LeadTimeInBusinessDays) AS MinLeadTime,
  MAX(o.LeadTimeInBusinessDays) AS MaxLeadTime
FROM publish_orders o
JOIN publish_product p ON o.ProductID = p.ProductID
WHERE p.ProductCategoryName IS NOT NULL 
  AND o.LeadTimeInBusinessDays IS NOT NULL
GROUP BY p.ProductCategoryName
ORDER BY p.ProductCategoryName;

ProductCategoryName AvgLeadTimeInBusinessDays TotalOrders MinLeadTime MaxLeadTime Accessories 5.01 13021 5 7 Bikes 5.01 12457 5 26 Clothing 5.01 23880 5 6 Components 5.01 34306 5 26

### Conclusion
Analytical Conclusions from the Case Study
1. Color Performance Analysis
Key Trends:

Market Shift: Clear consumer preference transition from Red (2021) to Black (2022-2023)
Black Dominance: Black generated $29M over two years, suggesting premium product positioning or mass appeal
Yellow Emergence: 2024 shows Yellow as the new leader, indicating possible product innovation or seasonal trend

Business Insights:

Black products are the company's cash cows - consistent high performers
Need to investigate the 50% revenue drop from Red (2021) to understand market dynamics
Yellow's sudden rise warrants deeper analysis - new product line? Marketing campaign effect?

2. Operational Efficiency Analysis
Remarkably Consistent Lead Times:

5.01 business days average across ALL categories
Minimal variance (mostly 5-7 days)
Notable outliers: Bikes and Components reaching 26 days

Strategic Implications:

Operational Excellence: Standardized fulfillment process across product lines
No Category Bias: Equal service levels regardless of product type
Hidden Issues: 26-day outliers suggest potential supply chain disruptions for complex products

3. Business Health Indicators
Positive Signals:

Consistent operational performance
Strong revenue growth (2021: $6M → 2023: $15M)
Successful product transitions

Areas of Concern:

Dramatic color preference shifts suggest volatile market
Limited product categories reporting (46 products uncategorized)
Potential supply chain vulnerabilities in complex products

### Appendices

#### Appendix A: Complete Code Repository
All code is available at: github.com/Bibidev42/Upstart13

#### Appendix B: Sample Data

**Products Sample:**
ProductID | ProductDesc | Color | ListPrice
680 | HL Road Frame - Black, 58 | Black | 1431.50
706 | HL Road Frame - Red, 58 | Red | 1431.50

**Orders Sample:**
SalesOrderID | OrderDate | ProductID | OrderQty | TotalLineExtendedPrice
43659 | 2021-05-31 | 776 | 1 | 2024.99

#### Appendix C: Business Glossary

- **Lead Time**: Business days between order and shipment
- **Extended Price**: Quantity × (Unit Price - Discount)  
- **Medallion Architecture**: Bronze (raw) → Silver (clean) → Gold (business-ready)
- **Business Days**: Monday-Friday, excluding weekends